File: backend/app/schemas/spo.py
import uuid
from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field

from app.models.spo import (
    ProcurementMethod,
    SPOAmendmentField,
    SPOAmendmentStatus,
    SPOStatus,
)


class SPOItemBase(BaseModel):
    product_id: uuid.UUID
    internal_sku: Optional[str] = None
    supplier_sku: Optional[str] = None
    description: str = Field(..., max_length=255)
    uom_id: uuid.UUID
    quantity_ordered: Decimal = Field(..., ge=0)
    unit_price: Decimal = Field(..., ge=0)
    discount_percent: Decimal = Field(Decimal("0"), ge=0)
    vat_rate: Decimal = Field(Decimal("0"), ge=0)
    expected_delivery_date: Optional[date] = None


class SPOItemCreate(SPOItemBase):
    rfq_award_line_id: Optional[uuid.UUID] = None
    procurement_request_item_id: Optional[uuid.UUID] = None
    line_number: int


class SPOItemResponse(SPOItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    spo_id: uuid.UUID
    line_number: int
    quantity_confirmed: Decimal
    quantity_backordered: Decimal
    quantity_received: Decimal
    quantity_accepted: Decimal
    quantity_damaged_rejected: Decimal
    quantity_invoiced: Decimal
    quantity_cancelled: Decimal
    open_quantity: Decimal
    vat_amount: Decimal
    total_price: Decimal
    price_amendment_pending: bool


class SPOCreate(BaseModel):
    supplier_id: uuid.UUID
    rfq_id: Optional[uuid.UUID] = None
    procurement_request_id: Optional[uuid.UUID] = None
    procurement_method: ProcurementMethod
    single_source_justification: Optional[str] = None
    supplier_reference: Optional[str] = None
    po_date: Optional[date] = None
    expected_delivery_date: Optional[date] = None
    warehouse_id: uuid.UUID
    currency: str = "AED"
    payment_terms_days: int = 0
    delivery_terms: Optional[str] = None
    items: List[SPOItemCreate]


class SPOAmendmentLineResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    amendment_id: uuid.UUID
    spo_item_id: uuid.UUID
    field_name: SPOAmendmentField
    old_value: Optional[str]
    new_value: Optional[str]


class SPOAmendmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    spo_id: uuid.UUID
    amendment_number: int
    status: SPOAmendmentStatus
    reason: str
    requires_supplier_reconfirmation: bool
    amended_by: Optional[uuid.UUID]
    approved_by: Optional[uuid.UUID]
    created_at: datetime
    applied_at: Optional[datetime]
    lines: List[SPOAmendmentLineResponse] = []


class SPOResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    supplier_id: uuid.UUID
    spo_number: str
    rfq_id: Optional[uuid.UUID]
    procurement_request_id: Optional[uuid.UUID]
    procurement_method: ProcurementMethod
    single_source_justification: Optional[str]
    supplier_reference: Optional[str]
    status: SPOStatus
    po_date: Optional[date]
    expected_delivery_date: Optional[date]
    warehouse_id: uuid.UUID
    currency: str
    payment_terms_days: int
    delivery_terms: Optional[str]
    subtotal: Decimal
    vat_amount: Decimal
    total_amount: Decimal
    quantity_ordered_total: Decimal
    quantity_confirmed_total: Decimal
    quantity_backordered_total: Decimal
    has_open_amendment: bool
    approved_by: Optional[uuid.UUID]
    approved_at: Optional[datetime]
    sent_at: Optional[datetime]
    acknowledged_at: Optional[datetime]
    closed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    items: List[SPOItemResponse] = []
    amendments: List[SPOAmendmentResponse] = []


class SPOAmendmentLineCreate(BaseModel):
    spo_item_id: uuid.UUID
    field_name: SPOAmendmentField
    # Canonical string form of the proposed value: Decimal string for
    # QUANTITY_ORDERED / UNIT_PRICE, ISO date for DELIVERY_DATE, raw text
    # for DELIVERY_TERMS / OTHER.
    new_value: str = Field(..., min_length=1, max_length=255)


class SPOAmendmentCreate(BaseModel):
    reason: str = Field(..., min_length=10)
    lines: List[SPOAmendmentLineCreate] = Field(..., min_length=1)


class SPOAmendmentApplyRequest(BaseModel):
    # Attestation that the supplier reconfirmed the amended terms out of band.
    # Required (422) when the amendment requires supplier reconfirmation.
    supplier_reconfirmed: bool = False


class SPOAcknowledgeLine(BaseModel):
    quantity_confirmed: Decimal = Field(..., ge=0)
    unit_price: Decimal = Field(..., ge=0)


class SPOAcknowledgeReq(BaseModel):
    lines: dict[uuid.UUID, SPOAcknowledgeLine]


class SPODeliveryScheduleCreate(BaseModel):
    tranche_number: int
    scheduled_quantity: Decimal = Field(..., gt=0)
    scheduled_date: date

File: frontend/src/api/spo.ts
import { apiClient } from './client';

export interface SPOItem {
  id: string;
  spo_id: string;
  line_number: number;
  product_id: string;
  internal_sku: string | null;
  supplier_sku: string | null;
  description: string;
  uom_id: string;
  quantity_ordered: number;
  quantity_confirmed: number;
  quantity_backordered: number;
  quantity_received: number;
  quantity_accepted: number;
  quantity_damaged_rejected: number;
  quantity_invoiced: number;
  quantity_cancelled: number;
  open_quantity: number;
  unit_price: number;
  discount_percent: number;
  vat_rate: number;
  vat_amount: number;
  total_price: number;
  expected_delivery_date: string | null;
  price_amendment_pending: boolean;
}

export interface SPO {
  id: string;
  workspace_id: string;
  supplier_id: string;
  spo_number: string;
  rfq_id: string | null;
  procurement_request_id: string | null;
  procurement_method: string;
  single_source_justification: string | null;
  supplier_reference: string | null;
  status: string;
  po_date: string | null;
  expected_delivery_date: string | null;
  warehouse_id: string;
  currency: string;
  payment_terms_days: number;
  delivery_terms: string | null;
  subtotal: number;
  vat_amount: number;
  total_amount: number;
  quantity_ordered_total: number;
  quantity_confirmed_total: number;
  quantity_backordered_total: number;
  has_open_amendment: boolean;
  approved_by: string | null;
  approved_at: string | null;
  sent_at: string | null;
  acknowledged_at: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
  items: SPOItem[];
}

export const getSPOs = async (workspaceId?: string): Promise<SPO[]> => {
  const params = workspaceId ? { workspace_id: workspaceId } : {};
  const response = await apiClient.get<SPO[]>('/api/v1/spos', { params });
  // Some endpoints might return SuccessResponse wrapped, but usually FastAPI returns the model directly unless wrapped.
  // Based on the schema provided, the backend returns SPOResponse (which is just the object) directly, not wrapped.
  // We'll return response.data directly assuming it's an array for list.
  return Array.isArray(response.data) ? response.data : (response.data as any).data;
};

export const getSPO = async (id: string): Promise<SPO> => {
  const response = await apiClient.get<SPO>(`/api/v1/spos/${id}`);
  return response.data;
};

export const createSPO = async (workspaceId: string, data: any): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/?workspace_id=${workspaceId}`, data);
  return response.data;
};

export const submitSPO = async (id: string): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/${id}/submit-approval`);
  return response.data;
};

export const approveSPO = async (id: string): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/${id}/approve`);
  return response.data;
};

export const sendSPO = async (id: string): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/${id}/send`);
  return response.data;
};

export const acknowledgeSPO = async (id: string, data: any): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/${id}/items/acknowledge`, data);
  return response.data;
};

export const cancelSPO = async (id: string, reason: string): Promise<SPO> => {
  const response = await apiClient.post<SPO>(`/api/v1/spos/${id}/cancel?reason=${encodeURIComponent(reason)}`);
  return response.data;
};
